### PR TITLE
Support variable length sequences in Seq2Seq example.

### PR DESCRIPTION
**Pull request by adarob**

### CODE REVIEW

**Comment by avital** [comment]
Just a drive-by review without looking closely but let's think carefully whether we want this in the base example as it seems to add quite a bit of code...

**Comment by adarob** [comment]
One option to simplify things is to move some of the support for variable length sequences into recurrent.py. 

There are 2 easy ways to do this: 

1. Add a function like the `hk.recurrent.dynamic_unroll` in haiku, but with length support for copying the state over once the full length is reached.
1. Add an RNN module whose `apply` takes care of the unrolling.

**Comment by adarob** [comment]
I'd also mention that most of the line changes are actually adding documentation and tests. I think the core changes aren't that large.

**Comment by marcvanzee** [comment]
I think adding this to the base example is useful since handling variable sequence lengths is a very basic feature of LSTMs. It does add a bit of code, but as Adam says as well, there's a lot of documentation and tests added.

I think adding general support for scanning over variable lengths is useful, but beyond the scope of this example.

**Comment by adarob** [comment]
I have an idea how to simplify things a bit as well. I'll send a patch
later today.

**Comment by adarob** [comment]
FYI, this PR demonstrates how simple the HOWTO is to add scheduled sampling after these changes: https://github.com/adarob/flax/pull/1

File: examples/seq2seq/train.py [hunk 1]
```diff
@@ -105,72 +130,140 @@ def decode_onehot(batch_inputs):
...
+        axis=1)
+    return final_state
 
 
 class Decoder(nn.Module):
   """LSTM decoder."""
 
-  def apply(self, carry, inputs):
+  def apply(self, init_state, inputs, teacher_force=False,
+            sample_temperature=1.0):
```
Comment (marcvanzee) [inline examples/seq2seq/train.py]: Is it necessary to have the sample_temperature?

Right now it seems we aren't using it (it is always 1), and it also isn't directly clear to me from the code how it should be used.

File: examples/seq2seq/train.py [hunk 2]
```diff
@@ -56,21 +56,39 @@
...
+
   def __init__(self, chars):
     self._chars = sorted(set(chars))
-    self._char_indices = dict((ch, idx) for idx, ch in enumerate(self._chars))
-    self._indices_char = dict((idx, ch) for idx, ch in enumerate(self._chars))
+    self._char_indices = dict(
+        (ch, idx + 2) for idx, ch in enumerate(self._chars))
+    self._indices_char = dict(
+        (idx + 2, ch) for idx, ch in enumerate(self._chars))
+    self._indices_char[self.pad_id] = '_'
```
[the remark this answers is not in the document]
Comment (adarob) [reply examples/seq2seq/train.py]: `?` is more commonly used for unk, but we could use it here if you'd prefer. We could also do `<pad>`, which is common. Let me know your preference.

File: examples/seq2seq/train.py (hunk 1 above)
Comment (adarob) [reply examples/seq2seq/train.py]: It's useful for being able to switch between "greedy" mode, sampling, etc. But, I agree that it's fine to remove since we're not using it here.

File: examples/seq2seq/train.py [hunk 3]
```diff
@@ -105,72 +130,140 @@ def decode_onehot(batch_inputs):
   return np.array(list(map(decode_inputs, batch_inputs)))
 
 
+def get_sequence_lengths(sequence_batch, eos_id=1):
+  """Returns the length of each one-hot sequence, including the EOS token."""
+  eos_row = sequence_batch[:, :, eos_id]
+  eos_idx = jnp.argmax(eos_row, axis=-1)  # returns first occurence
+  return jnp.where(
+      eos_row[jnp.arange(eos_row.shape[0]), eos_idx],
```
[the remark this answers is not in the document]
Comment (adarob) [reply examples/seq2seq/train.py]: No, this is to test for cases where eos_idx is 0 because the eos_id is not present. I will add a comment to make this more clear.

File: examples/seq2seq/train.py [hunk 4]
```diff
@@ -105,72 +130,140 @@ def decode_onehot(batch_inputs):
...
+      eos_idx + 1,
+      sequence_batch.shape[1]  # if there is no EOS, use full length
+  )
+
+
+def mask_sequences(sequence_batch, lengths):
```
[the remark this answers is not in the document]
Comment (adarob) [reply examples/seq2seq/train.py]: That's correct.

File: examples/seq2seq/train.py [hunk 5]
```diff
@@ -105,72 +130,140 @@ def decode_onehot(batch_inputs):
...
-    return carry
+    lengths = get_sequence_lengths(inputs, eos_id=eos_id)
+
+    lstm = nn.LSTMCell.partial(name='lstm')
+    init_lstm_state = nn.LSTMCell.initialize_carry(
+        nn.make_rng(),
+        (batch_size,),
+        hidden_size)
+
+    def encode(carry, x):
```
[the remark this answers is not in the document]
Comment (adarob) [reply examples/seq2seq/train.py]: I actually rewrote this to be more readable. Let me know what you think.

File: examples/seq2seq/train.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (marcvanzee) [reply examples/seq2seq/train.py]: I don't have a strong preference, I guess I was thinking of ' ' since I used that when padding (because I was printing it), but given that we won't use it anyway, '_' is fine to me.

**Comment by marcvanzee** [comment]
@googlebot I fixed it.

**Comment by marcvanzee** [comment]
I fixed it

File: examples/seq2seq/train.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcvanzee) [reply examples/seq2seq/train.py]: But if eos_id is not present, then eos_idx will be zero, right? (argmax([0, ..., 0])=0). So it seems to me that the max length is then returned correctly.

[the remark this answers is not in the document]
Comment (adarob) [reply examples/seq2seq/train.py]: I also want to handle the case that there is an EOS token at position 0, and therefore the length should be 1.

[the remark this answers is not in the document]
Comment (marcvanzee) [reply examples/seq2seq/train.py]: Ahh ok, I didn't think of that case.
